File: agent-sdk/python/examples/06-subagent.py
# 06 · Sub-agent delegation — an agent that prompts another agent, traced.
#
# New rung: 05-tools gave the agent a microservice-backed tool; here the
# capability is another AGENT. The coordinator forwards each prompt to a worker
# agent and streams the worker's answer back — a minimal multi-agent tree. The
# worker shares this process only to keep the demo self-contained.
#
# Observability: the service running each prompt mints a prompt_id and
# announces it on the trace subject (default afo.threads). The
# coordinator's handler runs with its trace bound as ambient context, so the
# nested worker.prompt(...) forwards parent/root automatically; tool_call_id=
# labels the edge with the model tool call being served. Watch the tree form:
#
#   nats sub afo.threads
#   uv run python examples/06-subagent.py --url nats://127.0.0.1:4222
#   nats req agents.prompt.coordinator.<owner>.main "hello" --replies=0 --reply-timeout=10s
#
# Connection: --context / --url, else $NATS_CONTEXT / $NATS_URL, else the
# selected `nats` context. Run with -h for the full flag list.

from __future__ import annotations

import argparse
import asyncio
import signal
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from synadia_ai.agents import Agents, DiscoverFilter, Envelope, ResponseChunk

from examples._connect_cli import (
    add_agent_identity_flags,
    add_connection_flags,
    connect_from_cli,
)
from synadia_ai.agent_service import AgentService, PromptStream


async def main() -> None:
    parser = argparse.ArgumentParser(
        description="Coordinator agent — delegates every prompt to a worker agent."
    )
    add_connection_flags(parser)
    add_agent_identity_flags(parser, agent="coordinator")
    args = parser.parse_args()

    nc = await connect_from_cli(args)

    # The worker: a stand-in for any other agent on the network.
    worker = AgentService(
        agent="worker",
        owner=args.owner,
        session_name="worker",
        nc=nc,
        description="worker agent — shouts the prompt back",
        heartbeat_interval_s=args.heartbeat_interval,
    )

    async def worker_handler(envelope: Envelope, stream: PromptStream) -> None:
        # A real worker would pass stream.trace_headers() to its model client.
        print(f"  worker    prompt={stream.prompt_id} parent={stream.trace.parent_prompt_id}")
        await stream.send(envelope.prompt.upper())

    worker.on_prompt(worker_handler)

    # One Agents client for the coordinator's outbound prompts.
    agents = Agents(nc=nc)
    await agents.start_tracking()

    coordinator = AgentService(
        agent="coordinator",
        owner=args.owner,
        session_name=args.session_name,
        nc=nc,
        description="coordinator agent — delegates every prompt to the worker",
        heartbeat_interval_s=args.heartbeat_interval,
    )

    async def coordinator_handler(envelope: Envelope, stream: PromptStream) -> None:
        print(f"coordinator prompt={stream.prompt_id} root={stream.root_id}")
        # Demo simplicity: re-discover the worker on every prompt. Production
        # handlers discover once and cache the Agent handle.
        found = await agents.discover(
            filter=DiscoverFilter(agent="worker", owner=args.owner), timeout=1.0
        )
        if not found:
            raise RuntimeError("worker agent not found — did its registration fail?")

        # Spawn the worker as a child execution — zero plumbing: the ambient
        # trace forwards parent/root; tool_call_id labels the edge as the tool
        # invocation it serves (omit it for a programmatic spawn).
        async for msg in found[0].prompt(envelope.prompt, tool_call_id="delegate-1"):
            if isinstance(msg, ResponseChunk):
                await stream.send(f"worker says: {msg.text}")

    coordinator.on_prompt(coordinator_handler)

    await worker.start()
    await coordinator.start()
    print(f"coordinator listening on {coordinator.subject.prompt}")
    print(f"worker listening on {worker.subject.prompt}")
    print("press Ctrl+C to stop")

    loop = asyncio.get_running_loop()
    stop = asyncio.Event()
    for _sig in (signal.SIGINT, signal.SIGTERM):
        loop.add_signal_handler(_sig, stop.set)
    try:
        await stop.wait()
    finally:
        print("\nshutting down…")
        await coordinator.stop()
        await worker.stop()
        await agents.close()
        await nc.close()


if __name__ == "__main__":
    asyncio.run(main())

File: agent-sdk/typescript/examples/06-subagent.ts
// 06 · Sub-agent delegation — an agent that prompts another agent, with trace edges.
//
// New rung on the ladder: 05-tools gave the agent a microservice-backed tool;
// here the capability is another AGENT. The coordinator forwards each prompt
// to a worker agent and streams the worker's answer back — a minimal
// multi-agent tree. The worker shares this process only to keep the demo
// self-contained.
//
// Observability: every prompt execution (a "thread") has a derived id, and
// the delegation is a parent→child edge an observing proxy reconstructs into
// an activity tree. The SDK binds the trace context around the handler, so
// prompting the worker inside toolScope(...) joins the tree and records the
// labeled edge with zero explicit plumbing. (The explicit equivalent —
// prompt(text, { trace: response.childTrace() }) + response.recordSpawn(...) —
// remains available outside a handler's context.)
//
// Try it — run this file, then prompt the coordinator (name "main"; the
// worker serves name "worker") from client-sdk/typescript/examples:
//
//   npx tsx examples/06-subagent.ts
//   npx tsx examples/02-prompt-text.ts "hello"
//
// Connection resolution: $NATS_CONTEXT > $NATS_URL > nats://127.0.0.1:4222.

import { connect as natsConnect } from "@nats-io/transport-node";
import { Agents, loadContextOptions, parseNatsUrl, toolScope } from "@synadia-ai/agents";
import { AgentService } from "@synadia-ai/agent-service";

async function main(): Promise<void> {
  const opts = process.env["NATS_CONTEXT"]
    ? await loadContextOptions(process.env["NATS_CONTEXT"])
    : process.env["NATS_URL"]
      ? parseNatsUrl(process.env["NATS_URL"])
      : { servers: "nats://127.0.0.1:4222" };
  const nc = await natsConnect(opts);

  const owner =
    process.env["SYNADIA_COORDINATOR_OWNER"] ??
    process.env["SYNADIA_OWNER"] ??
    process.env["NATS_AGENT_OWNER"] ??
    process.env["USER"] ??
    "anon";

  // The worker: a stand-in for any other agent on the network. It shares
  // this process (and connection) only to keep the demo self-contained.
  const workerSvc = new AgentService({
    nc,
    agent: "worker",
    owner,
    name: "worker",
    description: "worker agent — shouts the prompt back",
  });
  workerSvc.onPrompt(async (envelope, response) => {
    await response.send(envelope.prompt.toUpperCase());
  });

  // One Agents client for the coordinator's outbound prompts.
  const agents = new Agents({ nc });
  await agents.startTracking();

  const coordinator = new AgentService({
    nc,
    agent: "coordinator",
    owner,
    name: process.env["SYNADIA_NAME"] ?? process.env["NATS_AGENT_NAME"] ?? "main",
    description: "coordinator agent — delegates every prompt to the worker",
  });

  coordinator.onPrompt(async (envelope, response) => {
    // Demo simplicity: re-discover the worker on every prompt (no startup
    // ordering, survives worker restarts). Production handlers should
    // discover once and cache the Agent handle.
    const found = await agents.discover({ filter: { agent: "worker", owner } });
    if (found.length === 0) {
      throw new Error("worker agent not found — did its registration fail?");
    }
    const worker = found[0]!;

    // Spawn the worker as a child thread — zero plumbing: the ambient trace
    // bound around this handler forwards the tree root and records the
    // parent→child edge; toolScope() labels the edge as the tool invocation
    // it serves (without it: `programmatic`).
    const stream = await toolScope("delegate-1", () => worker.prompt(envelope.prompt));
    console.log(`delegating to worker thread ${stream.threadId}`);
    console.log(`  spawn marker: ${JSON.stringify(stream.spawnMarkerHeaders)}`);

    for await (const msg of stream) {
      if (msg.type === "response") {
        await response.send(`worker says: ${msg.text}`);
      }
    }
  });

  await workerSvc.start();
  await coordinator.start();
  console.log(`coordinator listening on ${coordinator.subject.prompt}`);
  console.log(`worker listening on ${workerSvc.subject.prompt}`);
  console.log("press Ctrl+C to stop");

  const shutdown = async (): Promise<void> => {
    console.log("\nshutting down…");
    await coordinator.stop();
    await workerSvc.stop();
    await agents.close();
    await nc.close();
    process.exit(0);
  };
  process.on("SIGINT", () => void shutdown());
  process.on("SIGTERM", () => void shutdown());
}

void main().catch((err: unknown) => {
  console.error("subagent demo failed:", err);
  process.exit(1);
});
